src/nvim/eval.c

```diff
@@ -9563,7 +9563,8 @@ static void f_getchar(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
     if (argvars[0].v_type == VAR_UNKNOWN) {
       // getchar(): blocking wait.
-      if (!(char_avail() || using_script() || input_available())) {
+      if (!(char_avail() || using_script() || input_available()
+            || in_callback)) {
```

Comment:
Will this not break waiting for a char? Only `input_enable_events()`  and `multiqueue_process_events()` should be guarded to disable deferred events.